### ISSUE TITLE
2D rendering bug in Pokémon - Mystery Dungeon - Explorers of Sky (also present in no$gba, formerly present in desmume)

### ISSUE THREAD

**Issue by wyatt8740** [StapleButter/melonDS#299]
Hi. I reported a bug in Desmume a while back where this particular scene in Pokémon: Mystery Dungeon: Explorers of Sky would render improperly. I just tested it today, and it also impacts your emulator, as well as No$GBA. Here's what it looks like:
![2dbug-melonds](https://user-images.githubusercontent.com/2400933/49912011-98dc0a00-fe56-11e8-9c4c-e0a73af8faa1.png)

Here's a save file from _immediately_ before it happens.
[PMD-sky-2dbug.sav.zip](https://github.com/StapleButter/melonDS/files/2674329/PMD-sky-2dbug.sav.zip)

Here's the bug report from DeSmuMe:
https://github.com/TASVideos/desmume/issues/92
Where Zeromus took a picture of the scene on a fat DS:
![Zeromus' picture of a real DS](https://user-images.githubusercontent.com/102701/29481816-7918685c-844b-11e7-921b-726e59ed881e.jpeg)

And a description of how they think the bug worked:
https://github.com/TASVideos/desmume/issues/92#issuecomment-323493310

And the commit that fixed it:
https://github.com/TASVideos/desmume/commit/48f0f69929d13d9ee1d2bf826f06fd31f1b2bdbf

**Comment by StapleButter** [comment]
this is interesting.

I have a quick idea how to fix this (basically swapping around some if statements in the special effect code), but I'd like to test this against hardware before (if semi-transparent OBJs have priority over the window color-effect flags, does that also apply to 3D-layer blending?)

**Comment by StapleButter** [comment]
'd like to say, thanks for all the notes on this, btw

**Comment by wyatt8740** [comment]
Not a problem. Were they helpful and relatively accurate?
I'm sort of at that stage where I feel like I essentially understand at some abstract level what the hardware does, but at which I still get bogged down by implementation details.

**Comment by StapleButter** [comment]
yeah, it was! saved a whole lotta time there. otherwise I'd have had to go down the standard path of examining what is displayed, the modes/blending/window/etc and why it's not rendering properly, etc

also, nope, don't know how to contact Martin :/

I know one of my fixes somehow made it to him tho (the bit where 3D matrix push/pop/store/restore in mode1 works same as in mode2), but dunno how. and this is another thing that was independently figured out and fixed in DeSmuME ages ago.
